Feature/dcc 5774 dcc repository investigate indexing song supported repositories

KevinFHartmann opened the pull request:
This is ready for a PR, assuming that the output is what we want it to be. :-)

It looks pretty good to me, but I might have missed/misunderstood something, so it would probably be good to have someone else review it, too. :-)

KevinFHartmann committed:
Enhancement: Added support for indexing from a SONG server.

KevinFHartmann committed:
Bugfix: Forgot to check in the pom.xml

KevinFHartmann committed:
Bugfix: Fixed broken pom.xml
Code Cleanup: Moved Json files into test
Enhancement: In SONGImporterTest, use MockSongClient so we don't need to set up a song server to run the tests.
Enhancement: Added the json file for studies that MockSongClient needs.
Code Cleanup: Removed unused imports, fixed up line spacing.

KevinFHartmann committed:
Change: Added the new feature configurations to the application.yml, so that it's obvious for how to set them
Change: Made the tests that depend on song be ignored by Maven, and the one that doesn't to get run by Maven.

KevinFHartmann committed:
Enhancement: SongClientTest now tests against multiple studies.
Enhancement: SongClientTest now actually tests the real functionality with mock data
Code Cleanup: Removed AbstractSongClient, renamed RealSongClient to SongClient, and made MockSongClient inherit it's functionality from SongClient.
Bugfix: Changed SongUri field to use a URL instead.
Code Cleanup: Cleaned up the SongImporter and SongProcessor code

KevinFHartmann committed:
Bugfix: Multiple projects need their own data, or it just looks like the data is being copied multiple times. Now MockSongClient uses the project name as the directory name to get it's data from.
Enhancement: Implemented experiments, and hopefully mapped the data in them to the correct things.
Bugfix: The configuration still wasn't handling Song URLs properly. Fixed.

KevinFHartmann committed:
Copyright: Added missing copyright headers to dcc-repository-song

KevinFHartmann committed:
Code Cleanup: Formatting and Copyrights
Bugfix: Restored list of importers to test back to normal in RepositoryImporterTest.java

andricDu commented on dcc-repository-song/src/main/java/org/icgc/dcc/repository/song/core/SongProcessor.java:
```diff
@@ -0,0 +1,217 @@
+    private final RepositoryFileContext context;
+
+    public Iterable<RepositoryFile> getRepositoryFiles(Iterable<SongAnalysis> analyses) {
+        return StreamSupport.stream(analyses.spliterator(), false).
```
Method reference and static imports

```java
        return stream(analyses.spliterator(), false).
                flatMap(this::convertFiles).collect(toList());
```

andricDu commented on dcc-repository-song/src/main/java/org/icgc/dcc/repository/song/core/SongProcessor.java:
```diff
@@ -0,0 +1,217 @@
+    public Stream<RepositoryFile> convertFiles(SongAnalysis analysis) {
+        val results = new ArrayList<RepositoryFile>();
+
+        for(val file: analysis.getFiles()) {
```
This whole method could be shortened down to:

```java
        return analysis.getFiles().stream()
            .filter(f -> !isIndexFile(f.get(fileName)))
            .map(f -> convert(f, analysis));
```

andricDu reviewed (changes requested):
I've noticed some of the files are using 4 spaces for indentation. Lets bring them back to 2 for consistency with the rest of our code base. 

Once you set the default in IntelliJ you can just use the IDE to reformat.

KevinFHartmann committed:
Code cleanup: Made the fixes in the PR to SongProcessor.java
Enhancement: Added a "songToken" setting in the config file, together with changes to the client code to access the SONG server using the token, if any is set.

KevinFHartmann committed:
Bugfix: The PCAWG processor was applying post-processing fixes after converting
everything, but the SONG processor wasn't. Made the SONG processor apply the same post processing steps,
Code Cleanup: Refactored the convert() method in SongProcessor into functions
Bugfix: Now correctly handle several other data fields that weren't being filled in correctly.

KevinFHartmann committed:
Bugfix: Added the primary-source field
Code cleanup: Deleted unused code.

KevinFHartmann committed:
Enhancement: Checked in file with javascript functions for integration testing dcc-repository-song against production, with usage comments at the top.

KevinFHartmann committed:
Bugfix: The project name should be "PCAWG", not the sub-project name, to match what we have
now in Collaboratory.
Bugfix: We still want SONG to use the existing Collaboratory download URL, not it's own URL.

KevinFHartmann commented:
This is ready to be reviewed again. :-)

KevinFHartmann committed:
Bugfix: We now set the id fields the same way PCAWG does.

KevinFHartmann committed:
Enhancement: We now expect the SONG server to send us a field called "fileAccess" for each file in an analysis: which we use to set the access level for the data file.
Enhancement: We now assume the index file is the main file name with the index file type added as an extension, rather than choosing the first index file in the analysis.
Enhancement: We assume the metadata filename is the repositories metadata path + the analysis id

KevinFHartmann committed:
Bugfix: Changed metadata lookup for SONG to give the object ID of the XML file,
not the analysis id.
Change: Changed name of song's code from song to collaboratory-song, since
we want to be able to index other song servers soon.

KevinFHartmann committed:
Change: Changed song repository code from "collaboratory-song" to song

KevinFHartmann committed:
Bugfix: No longer hard code collaboratory to verify SONG code works for collab
Change: Renamed code for collab to "collaboratory-old", so that the SONG source could use the name "collab".

KevinFHartmann committed:
Merge branch 'develop' into feature/DCC-5774-DCC-Repository-Investigate-indexing-SONG-supported-Repositories

KevinFHartmann committed:
Updated with 1.8 branch

KevinFHartmann committed:
This version now runs to completion on production data, even for EGA data.

KevinFHartmann committed:
Code Cleanup: Ran Code Format On Everything

KevinFHartmann committed:
Change: For SONG, we now categorize the study in the same way that GDC does.
Change: We now categorize the dataType for SequencingRead analysis files as
follows:
-------
Song has three categories of Sequencing Read files to categorize:
1. BAM fileType, which know is aligned (or unaligned) ⇒ '(Una/A)ligned Reads'
2. BAM fileType, we don't know whether it's aligned or not aligned (ie. aligned is null) ⇒ 'Sequencing Reads'
3. FASTQ file ⇒ 'Unaligned Reads'

We can detect the `fileType` from the fileType field in the 'file' section of the JSON, and we can get the 'aligned' field from the 'experiment' subsection of the Analysis section.
----------

KevinFHartmann committed:
Fixs for FASTQ and Jun Jun's other rules

KevinFHartmann committed:
Change: We now set the analysisType to the SONG analysis type, instead of hard coding "No Data"

KevinFHartmann committed:
Change: Renamed the repository COLLAB to COLLAB_OLD; renamed SONG to COLLAB
Bugfix: Annotated SongImporterTest with @Ignore, plus a message about it's dependencies

andricDu commented on dcc-repository-core/src/main/java/org/icgc/dcc/repository/core/util/RepositoryFiles.java:
```diff
@@ -40,7 +41,16 @@ public static boolean isPCAWGFile(@NonNull RepositoryFile file) {
 
   public static boolean isPCAWGFileCopy(@NonNull FileCopy fileCopy) {
-    return fileCopy.getRepoOrg().equals(PCAWG.getId());
+    if (fileCopy == null) {
```
fileCopy cannot be null due to the `@NonNull`

Commit:
Removed null check due to already existing @NonNull
